firedrake/solving_utils.py

```diff
@@ -229,10 +266,38 @@ def split(self, fields):
                         W = V.sub(field_renumbering[index])
                     if cmpt is not None:
                         W = W.sub(cmpt)
-                    bcs.append(type(bc)(W,
-                                        bc.function_arg,
-                                        bc.sub_domain,
-                                        method=bc.method))
+                    if isinstance(bc, DirichletBC):
+                        return DirichletBC(W,
+                                           bc.function_arg,
+                                           bc.sub_domain,
+                                           method=bc.method)
+                    elif isinstance(bc, EquationBC):
+                        bc_F = replace(splitter.split(bc.F, argument_indices=(field, )), {bc.u: u})
+                        bc_J = replace(splitter.split(bc.J, argument_indices=(field, field)), {bc.u: u})
+                        bc_Jp = None if bc.Jp is None else replace(splitter.split(bc.Jp, argument_indices=(field, field)), {bc.u: u})
+                        ebc = EquationBC(bc_F == 0,
+                                         subu,
+                                         bc.sub_domain,
+                                         method=bc.method,
+                                         bcs=None,
+                                         J=bc_J,
+                                         Jp=None if bc.Jp_eq_J else bc_Jp,
+                                         V=W,
+                                         is_linear=bc.is_linear)
+                        for bbc in bc.bcs:
+                            bc_temp = rcollect_bcs(bbc)
+                            # Due to the "if index", bc_temp can be None
+                            if bc_temp is not None:
+                                ebc.add(bc_temp)
```

Comment:
Again, it looks like simple iteration over the bcs:
```python
for bc in itertools.chain(*problem.bcs):
   ...
```
Would be preferable to this recursion.

Reply:
Or if not this, then we should think about how these kind of "tree-visitor" patterns can be encapsulated in easier to follow code.

Reply:
This part is also greatly simplified by `reconstruct` method.  I also changed the `reconstruct` method of `DirichletBC` class accordingly (this change should be backward compatible).